### PR TITLE
Support for macaroon root key rotation

### PR DESCRIPTION
**Pull request by mrwhythat**
Implements #286. Based off #313.
Subtasks from issue:
- [x] Update the `RootKeyStorage` implementation to generate a new root key and ID every time `RootKey()` is called.
- [x] Update `genMacaroons()` in `lnd.go` to add a `TimeBeforeCaveat` with a sane default timeout and possibly a command-line option in config.go to modify that timeout if desired.
- [x] Update the code in `lndMain()` in `lnd.go` that checks for macaroon file existence to also check for macaroon validity/expiration.

### CODE REVIEW

File: macaroons/constraints.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+// IPLockConstraint locks macaroon to a specific IP address.
+// If address is an empty string, this constraint does nothing to
+// accommodate default value's desired behavior.
+func IPLockConstraint(ipAddr string) func(*macaroon.Macaroon) error {
+	return func(mac *macaroon.Macaroon) error {
+		if ipAddr != "" {
+			macaroonIPAddr := net.ParseIP(ipAddr)
+			if macaroonIPAddr == nil {
+				return fmt.Errorf("incorrect macaroon IP-lock address")
```
Comment (jimpo) [inline macaroons/constraints.go]: Maybe worth including `ipAddr` in the error message?

File: macaroons/auth.go [hunk 2]
```diff
@@ -67,6 +72,16 @@ func ValidateMacaroon(ctx context.Context, method string,
...
 	}
 
+	// Get peer info and extract IP address from it for macaroon check
+	pr, ok := peer.FromContext(ctx)
+	if !ok {
+		return fmt.Errorf("unable to get peer info from context")
+	}
+	peerAddr, _, err := net.SplitHostPort(pr.Addr.String())
+	if err != nil {
+		return fmt.Errorf("unable to parse peer address")
```
Comment (jimpo) [inline macaroons/auth.go]: Maybe worth including `pr.Addr` in the error message?

**Comment by mrwhythat** [comment]
@jimpo , thanks for the review. Both of the error messages you mentioned belong to #313, which was approved by @aakselrod and is waiting to be merged in, so since this branch is based off that one, I think it would be better to wait for it to be merged and add changes you suggested afterwards.

**Comment by mrwhythat** [comment]
Rebased on master.

**Comment by mrwhythat** [comment]
Hey @aakselrod, any changes suggested?

**Comment by guggero** [comment]
This has been untouched for more than 2 years now. Not sure what to do with it.
In my opinion, adding an expiration date to each macaroon just creates another hassle for users of the API. I think what we really want is the ability to revoke a specific macaroon in conjunction with the bakery.
I'm in favor of closing this PR, together with the feature request (#286).

@Roasbeef, @wpaulino what do you think?

**Comment by halseth** [comment]
Better to support adding timeouts through the bakery?

**Comment by guggero** [comment]
We've discussed this internally and decided that we don't want to include this feature in lnd as is.
The following reasons were stated:
 - gRPC and REST users would run into an error every week with the default settings
 - lnd needs to be restarted for new macaroons to be created
 - the new [`bakemacaroon` command](https://github.com/lightningnetwork/lnd/tree/master/macaroons) allows users to set a timeout to a newly baked macaroon.
 - more fine-grained control over root keys should be implemented (see #3715).

**Pull request closed without merging**